### PR TITLE
Throw exception when query partition assignment in Maintenance mode

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#1854

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

API GetIdealAssignmentForFullAuto will return the potential partition assignment regardless of the cluster state. If the cluster is in maintenance mode, this API will returns potential assignment considering the input change.
However, GetImmediateAssignmentForWagedFullAuto will honor the cluster state. If the cluster is in maintenance mode, this API will use maintenance rebalancer.

### Tests

- [X] The following tests are written for this issue:

TestResourceAssignmentOptimizerAccessor

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java [hunk 1]
```diff
@@ -322,6 +322,13 @@ private void computeWagedAssignmentResult(List<IdealState> wagedResourceIdealSta
       InputFields inputFields, ClusterState clusterState, String clusterId,
       AssignmentResult result) {
 
+    // If the cluster is in M mode, throw an exception
```
Comment (jiajunwang) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java:325]: nit, M mode is not official and may be misunderstood. Please use full name.

**Review by jiajunwang — APPROVED** [review]
LGTM.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java (hunk 1 above)
Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java]: TFTR. Updated.

**Comment by xyuanlu** [comment]
This PR is ready to be merged. Approved by @jiajunwang 

Final commit message:
Throw exception for partitionAssignment when cluster in Maintenance mode.

**Pull request merged**